### DISCUSSION TITLE
Runtime configs in nextjs

### DISCUSSION THREAD

**Discussion by ederparaiso** [vercel/next.js#36711] [Help] [answered]
Hi, I'm trying to setup my nextjs app to use runtime configurations. Basically, I have an endpoint url that needs to be available trough docker env vars.
I configured following these docs  but it isn't working. My app still using default values from .env file. Could anyone help to understand what I missed or did wrong? 
Thanks!

docs:
https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration
https://nextjs.org/docs/advanced-features/custom-app

steps:
1- added to my next.config.js
```js
publicRuntimeConfig: {
  NEXT_PUBLIC_BACKEND_HOST: process.env.NEXT_PUBLIC_BACKEND_HOST,
},
```
2- retrieved config in my pages
```js
const { publicRuntimeConfig } = getConfig()
const baseURL = publicRuntimeConfig.NEXT_PUBLIC_BACKEND_HOST
```
3- created a custom app to setup getInitialProps
> Runtime configuration won't be available to any page (or component in a page) without getInitialProps.
```js
import App from 'next/app'

function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}
MyApp.getInitialProps = async (appContext) => {
  const appProps = await App.getInitialProps(appContext);
  return { ...appProps }
}
export default MyApp
```

**Comment by vadymshymko** [comment]
Hello!) According to the docs:
> Runtime configuration won't be available to any page (or component in a page) without getInitialProps.

So, maybe you forgot to add getInitialProps to your  page?

Also, it seems to me that there is a high probability that the value of your variable is overwritten just because of the presence of .env files, which are perceived by next as the most priority source.  Based on this, I think it makes sense to try to run the build in docker without an env file.

In addition, I recommend paying attention to what is said in the documentation:
> Generally you'll want to use [build-time environment variables](https://nextjs.org/docs/basic-features/environment-variables) to provide your configuration. The reason for this is that runtime configuration adds rendering / initialization overhead and is incompatible with [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization).

Hope this helps

**Reply by ederparaiso** [reply]
Thanks for your help! 
> So, maybe you forgot to add getInitialProps to your page?

I created a custom app as described here:
https://nextjs.org/docs/advanced-features/custom-app

> Also, it seems to me that there is a high probability that the value of your variable is overwritten just because of the presence of .env files,
...

Good observation. I removed the .env file but it's still not working. 

> In addition, I recommend paying attention to what is said in the documentation:

basically my requirements are: we have to provide an endpoint url read at runtime from our docker env configurations in a k8s cluster.
We created a example app with all current configuration to show the ideia. May I configured something still wrong?

https://github.com/rafaela2/hello-world-next

**Reply by vadymshymko** [reply]
Hello. Thanks for the example.
I understand why it does not work as you expect) The thing is that publicRuntimeConfig is filled with values ​​at the build stage, and not at runtime. That is why, if we do not delete the .env file, we see the text "Hello from .env" in the browser. And that's why we don't see any text in the browser after we delete the .env file. That is, in your case, using the current configuration of the docker run command, we will never see the text "Hello from docker " in the browser, because in fact we will expect to see the value of the runtime variable in the browser, which is impossible)
But, there is a solution. Moreover, there are 2 options:
1. Use [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props) on the page on which we want to display the value of any variable from process.env:
```
import Head from "next/head";
import styles from "../styles/Home.module.css";

function Home({ somePropName }) {
  return (
    <div className={styles.container}>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <p className={styles.description}>{somePropName}</p>
      </main>
    </div>
  );
}

export function getServerSideProps() {
  return {
    props: {
      somePropName: process.env.NEXT_PUBLIC_TITLE,
    }
  };
};

export default Home;

```
2. Do not use --env in docker run, but use --build-arg in docker build (precompile it):
`--build-arg NEXT_PUBLIC_TITLE=hello_from_docker`
then, in your Dockerfile (after `FROM node:16-alpine AS deps`):
```
ARG NEXT_PUBLIC_TITLE
ENV NEXT_PUBLIC_TITLE=${NEXT_PUBLIC_TITLE} 
```
in pages/index.tsx:
```
import Head from "next/head";
import styles from "../styles/Home.module.css";

function Home() {
  return (
    <div className={styles.container}>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <p className={styles.description}>{process.env.NEXT_PUBLIC_TITLE}</p>
      </main>
    </div>
  );
}
```

I would prefer the 2nd option, as it allows you to use [SSG](https://nextjs.org/docs/basic-features/data-fetching/get-static-props), which allows you to improve site performance

It's also worth noting that in both cases we can/should delete getInitialProps in _app (Adding a custom [getInitialProps](https://nextjs.org/docs/api-reference/data-fetching/get-initial-props) in your App will disable [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) in pages without [Static Generation](https://nextjs.org/docs/basic-features/data-fetching/get-static-props))

**Comment by ederparaiso** [comment, marked as answer]
Actually, my needs are the same as #17641 . 
I found this topic recently. I ended up using this approach to achieve my needs: https://raphaelpralat.medium.com/system-environment-variables-in-next-js-with-docker-1f0754e04cde
